### PR TITLE
✨(frontend) add component for dashboard's course run

### PR DESCRIPTION
**Pull request by NathanVss**
## Purpose

In the future dashboard users will be able to see course runs with various states. So we begin the implementations of sketches.

Capture d’écran 2022-09-15 à 09 07 17

## Todo

- [x] Basic skeleton
- [x] Responsive
- [x] Css theme colors
- [x] Ellipsis
- [x] Handle different objects like `Enrollment` and `Order`
- [x] Factory for mocks
- [x] Update css classnames

### CODE REVIEW

**Comment by sampaccoud** [comment]
@NathanVss would be great if you could paste a screenshot of the widget in the description of each such PR! :star_struck:

**Comment by NathanVss** [comment]
> @NathanVss would be great if you could paste a screenshot of the widget in the description of each such PR! 🤩

Good idea ! It's done :)

**Comment by sampaccoud** [comment]
:smiling_face_with_three_hearts:

**Comment by NathanVss** [comment]
@jbpenrath ready for an intermediate review, concerning models and global structure of the new components.

File: src/frontend/js/components/DashboardCourseRun/_styles.scss [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        width: 70px;
+        background-size: contain;
+        background-repeat: no-repeat;
+        background-position: center;
+        box-shadow: r-theme-val(course-glimpse, base-shadow);
+        border-radius: $border-radius;
+      }
+
+      &__title {
+        margin-top: 10px;
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardCourseRun/_styles.scss]: For spacing properties, please try to use `rem`

File: src/frontend/js/components/Button/index.tsx [hunk 2]
```diff
@@ -0,0 +1,9 @@
+interface ButtonProps {
```
Comment (jbpenrath) [inline src/frontend/js/components/Button/index.tsx:1]: I'll add story also for this generic component

File: src/frontend/js/components/Button/index.tsx [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+  label: string;
+  color: 'primary' | 'secondary' | 'outline-primary' | 'outline-secondary';
+}
+
+export const Button = (props: ButtonProps) => {
+  const classes = ['button', 'button--' + props.color];
+  return <button className={classes.join(' ')}>{props.label}</button>;
```
Comment (jbpenrath) [inline src/frontend/js/components/Button/index.tsx]: I'll use `children` instead of a `label` property.
```suggestion
  return <button className={classes.join(' ')}>{children}</button>;
```

File: src/frontend/js/components/DashboardCourseRun/index.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { CourseRun } from '../../types';
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardCourseRun/index.tsx]: You delete this component in the next commit. You could squash these two modifications to only show the final version of the component.

File: src/frontend/js/components/DashboardItem/_styles.scss [hunk 5]
```diff
@@ -31,22 +34,36 @@
...
+    position: relative;
+
+    &::before {
+      position: absolute;
+      left: -20px;
+      top: 20px;
+      content: '';
+      width: 10px;
+      height: 10px;
+      background: r-color(indianred3);
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/_styles.scss]: You should use `r-theme-val` as within site-factory we must be able to use another color.

File: src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  return {
+    canAccessCourse: enrollment.is_active && !isClosed,
+    isClosed,
+  };
+};
+
+export const DashboardItemEnrollment = ({ enrollment }: DashboardItemCourseRunProps) => {
+  const { course } = enrollment;
+  if (!course) {
+    throw new Error('CourseRun must provide course attribute');
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx:45]:
```suggestion
    throw new Error('Enrollment must provide course attribute');
```

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
+import { ComponentMeta, ComponentStory } from '@storybook/react';
+import { DashboardItemOrder } from './DashboardItemOrder';
+import { Order, OrderState } from '../../types/Joanie';
+import { enrollment } from './mock.stories';
+
+const order: Order = {
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemOrder.stories.tsx:6]: Could we use `OrderFactory` here ?

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.tsx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        <>
+          <div className="dashboard-course-run__block__status">
+            <Icon name="icon-school" />
+            <div>
+              For PR @jbpenrath: Don't know which data to use here, all course course_runs ? But
+              does it makes sense for a gradeo product ?
+            </div>
+          </div>
+          <Button
+            label={"@jbpenrath: How to get the course's url ?"}
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemOrder.tsx:34]: There will be a dashboard course detail view, I think with following url 
`/dashboard/courses/:course_code`

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.tsx [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    <DashboardItem
+      title={course.title}
+      headRef={'Ref. ' + course.code}
+      imageUrl={DEMO_IMAGE_URL}
+      footer={
+        <>
+          <div className="dashboard-course-run__block__status">
+            <Icon name="icon-school" />
+            <div>
+              For PR @jbpenrath: Don't know which data to use here, all course course_runs ? But
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemOrder.tsx]: I think all target_courses should be listed here. Then according if the user is enrolled or not to the course run, the layout should be different. I think we have to talk about that during a meeting.

File: src/frontend/js/components/DashboardItem/DashboardSubItemsList.tsx [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+interface DashboardSubItemsListProps {
+  subItems: ReactNode[];
+}
+
+export const DashboardSubItemsList = ({ subItems }: DashboardSubItemsListProps) => {
+  return (
+    <div className="dashboard-sub-course-run-list">
+      {subItems.map((subItem) => (
+        <div className="dashboard-sub-course-run-list__item">{subItem}</div>
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardSubItemsList.tsx]: do not forget to pass a `key` or you could use `React.Children.map`

File: src/frontend/js/components/Icon/index.tsx [hunk 11]
```diff
@@ -40,7 +40,8 @@ export type IconType =
   | 'icon-stopwatch'
   | 'icon-three-vertical-dots'
   | 'icon-twitter'
-  | 'icon-warning';
+  | 'icon-warning'
+  | 'icon-school';
```
Comment (jbpenrath) [inline src/frontend/js/components/Icon/index.tsx:44]: Please order alphabetically to help dev to find icon by name easily.

File: src/frontend/js/types/Joanie.ts [hunk 12]
```diff
@@ -122,6 +122,10 @@ export enum OrderState {
 export interface Order {
   id: string;
   course: string;
+  // I wanted to use the `course` attribute, but it is already in use for the course's id.
+  // Maybe would it makes more sens to rename `course` to `courseId` in order to use `course`
```
Comment (jbpenrath) [inline src/frontend/js/types/Joanie.ts:126]: If we need to have more information than id for the course, the api could return these information instead of a simple id.

`course: Course`

File: src/richie/apps/core/templates/richie/icons.html [hunk 13]
```diff
@@ -1,127 +1,165 @@
 <svg
-  style="position: absolute; width: 0; height: 0;"
-  width="0"
-  height="0"
-  version="1.1"
-  xmlns="http://www.w3.org/2000/svg"
-  xmlns:xlink="http://www.w3.org/1999/xlink"
+  style='position: absolute; width: 0; height: 0;'
```
Comment (jbpenrath) [inline src/richie/apps/core/templates/richie/icons.html:2]: Whoops, internal linter was here ?

**Review by jbpenrath — COMMENTED** [review]
Please reorganize your commits as some files are created within a commit then remove into the next one that's not help to review this PR.

File: src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    id: 'components.DashboardItemEnrollment.statusNotActive',
+    description: '',
+    defaultMessage: 'NOT ACTIVE',
+  },
+  accessCourse: {
+    id: 'components.DashboardItemEnrollment.gotoCourse',
+    description: '',
+    defaultMessage: 'ACCESS COURSE',
+  },
+});
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx:28]: Do not translate this content as this is static example waiting to have dynamic values.

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.stories.tsx (hunk 7 above)
Comment (NathanVss) [reply src/frontend/js/components/DashboardItem/DashboardItemOrder.stories.tsx:6]: I think so, planned to do that after this first overall review :)

File: src/frontend/js/types/Joanie.ts (hunk 12 above)
Comment (NathanVss) [reply src/frontend/js/types/Joanie.ts:126]: I introduced a new attribute `courseEntity?: Course;` because I was wondering that maybe changing the type of `course` from `string` to `Course` would break something somewhere. 

But indeed it seems that the `course` attribute is not used anywhere, could you confirm my thought ?

Comment (jbpenrath) [reply src/frontend/js/types/Joanie.ts:126]: Yes I confirm ! `order.course` is currently not used so you can change the type of this property.

File: src/frontend/js/utils/test/factories.ts [hunk 15]
```diff
@@ -32,6 +32,9 @@ export const EnrollmentFactory = createSpec({
   created_at: derived(() => faker.date.past()().toISOString()),
   user: faker.datatype.number(),
   course_run: faker.datatype.number(),
+  course: derived(() => CourseFactory.generate()),
+  is_active: true,
+  resource_link: faker.internet.url(),
```
Comment (jbpenrath) [inline src/frontend/js/utils/test/factories.ts:37]: You should use `JoanieEnrollmentFactory` instead. Currently we manage two types of enrollment. Those passed by Richie then those passed by Joanie and there are some differences between both so we have two factories ..

File: src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+export const DashboardItemEnrollment = ({ enrollment }: DashboardItemCourseRunProps) => {
+  const { course } = enrollment;
+  if (!course) {
+    throw new Error('Enrollment must provide course attribute');
+  }
+
+  return (
+    <DashboardItem
+      title={course.title}
+      headRef={'Ref. ' + course.code}
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx:51]: I tend to think that this component may be used for a totally different entity in the future where this data could be anything else than a "Ref ..." format, like a date, a username for instance. Not putting the `ref` semantic in the `DashboardItem` keeps it more generic in my POV. Nope ?

[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx:51]: Yep ok !
Then I think in real implementation there will be an intl message here to we do not have to do this concatenation.

**Comment by jbpenrath** [comment]
About responsive, I'll prevent the button to shrink and instead shrink the text on the left.

image

File: src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+  enrollment: Enrollment;
+}
+
+const useDashboardEnrollment = (enrollment: Enrollment) => {
+  const isClosed = new Date(enrollment.end) <= new Date();
+  return {
+    canAccessCourse: enrollment.is_active && !isClosed,
+    isClosed,
+  };
+};
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx:40]: The `Enrollment` type doesn't include this property: `export interface Enrollment extends Omit<CourseRun, 'state'> {`

:/

File: src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+  const { canAccessCourse } = useDashboardEnrollment(enrollment);
+  return (
+    <>
+      <div className="dashboard-item__block__status">
+        <Icon name="icon-school" />
+        <DashboardItemEnrollmentStatus enrollment={enrollment} />
+      </div>
+      <Button
+        color="outline-primary"
+        disabled={!canAccessCourse}
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx:69]: Need to talk about that in a meeting for me to understand well 🙏

File: src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/components/DashboardItem/DashboardItemEnrollment.tsx:51]: You're right :)

File: src/frontend/js/components/DashboardItem/index.tsx [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+        <header className="dashboard-item__block__head">
+          <img
+            data-testid="dashboard-item__block__head__thumbnail"
+            className="dashboard-item__block__head__thumbnail"
+            src={props.imageUrl}
+            alt=""
+          />
+          <div className="dashboard-item__block__head__captions">
+            <div className="dashboard-item__block__head__code">{props.code}</div>
+            <h5 className="dashboard-item__block__head__title">{props.title}</h5>
```
Comment (Leimi) [inline src/frontend/js/components/DashboardItem/index.tsx:26]: hey, hope it's OK for me to make some accessibility related comments :innocent: 

I'd argue here it'd be better if the heading appeared before the code in the DOM. As people with screen reader can navigate by jumping from heading to heading, if an information related to some heading appears in the DOM before it, it can be missed (since a screen reader then reads the content in the order of the DOM).

So the idea here would be to swap the `__code` and `__title` lines and rearrange them visually via flexbox + css `order` prop, for example.

Comment (NathanVss) [reply src/frontend/js/components/DashboardItem/index.tsx:25]: Hey, of course it's ok ! I just made the change. I will learn lots of things :)

File: src/frontend/js/components/DashboardItem/DashboardItemEnrollment.spec.tsx [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+
+describe('<DashboardItemEnrollment/>', () => {
+  const enrollment: Enrollment = JoanieEnrollmentFactory.seed(3).generate();
+  enrollment.course_run = {
+    ...enrollment.course_run,
+    end: '2039-10-01T01:23:37+00:00',
+    start: '2022-09-09T12:02:44+00:00',
+  };
+
+  const course = enrollment.course_run.course!;
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemEnrollment.spec.tsx:15]: We prefer to use full random values in order to detect edge cases.
I recommend you to create an enrollment for each tests.

You are able to generate a course run with a given state with random values. e.g:

```js
 const courseRun: CourseRun = JoanieCourseRunFactory.extend({
        // - Course Run opened for enrollment but not yet started
        ...cr,
        enrollment_start: faker.date.past(0.25)().toISOString(),
        enrollment_end: faker.date.future(0.5)().toISOString(),
        start: faker.date.future(0.75)().toISOString(),
        end: faker.date.future(1.0)().toISOString(),
        }).generate();

const enrollment = JoanieEnrollmentFactory.extend({ course_run: courseRun }).generate()
```

File: src/frontend/js/components/DashboardItem/DashboardItemEnrollment.stories.tsx [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+export const Closed = Template.bind({});
+Closed.args = {
+  enrollment: {
+    ...enrollment,
+    course_run: { ...enrollment.course_run, end: '2021-10-01T01:23:37+00:00' },
+  },
+};
+
+export const NotActive = Template.bind({});
+NotActive.args = { enrollment: { ...enrollment, is_active: false } };
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemEnrollment.stories.tsx:31]:
```suggestion
export const Inactive = Template.bind({});
```

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+    expect(button).toHaveAttribute('href', '/dashboard/courses/' + order.course!.code);
+  });
+
+  it('render component with enrollments', () => {
+    const enrollment: Enrollment = JoanieEnrollmentFactory.seed(3).generate();
+    enrollment.course_run = {
+      ...enrollment.course_run,
+      end: '2039-10-01T01:23:37+00:00',
+      start: '2022-09-09T12:02:44+00:00',
+    };
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx:28]: Ditto use random values for dates.

File: src/frontend/js/components/DashboardItem/DashboardItemEnrollmentFooter.tsx [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+      </Button>
+    </>
+  );
+};
+
+const DashboardItemEnrollmentStatus = ({ enrollment }: { enrollment: Enrollment }) => {
+  const intl = useIntl();
+
+  if (!enrollment.is_active) {
+    return <FormattedMessage {...messages.statusNotActive} />;
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/components/DashboardItem/DashboardItemEnrollmentFooter.tsx:51]: Ok I'm changing the text for now. But how can we end to this situation where a user has an `Enrollment` but is not enrolled ? 

When a admin cancels its enrollment ?

File: src/frontend/js/utils/test/factories.ts [hunk 24]
```diff
@@ -25,6 +25,7 @@ export const CourseRunFactory = createSpec({
   languages: faker.random.locale(),
   state: CourseStateFactory,
   starts_in_message: null,
+  course: derived(() => CourseFactory.generate()),
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/utils/test/factories.ts:28]: I can't because CourseFactory is declared after, but I will re-order all the factories to achieve that.

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+} from 'utils/test/factories';
+import { getByRole, getByText, render, screen } from '@testing-library/react';
+import { IntlProvider } from 'react-intl';
+import { DashboardItemOrder } from 'components/DashboardItem/DashboardItemOrder';
+import { Order } from 'types/Joanie';
+import faker from 'faker';
+import { FORMAT_DATE_OPTIONS } from 'utils/useDateFormat';
+
+describe('<DashboardItemOrder/>', () => {
+  it('render component without enrollments', () => {
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx:14]:
```suggestion
  it('renders component without enrollments', () => {
```

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+        <DashboardItemOrder order={order} />
+      </IntlProvider>,
+    );
+    screen.getByText(order.course!.title);
+    const button = screen.getByTestId('dashboard-item-order__button');
+    expect(button).toBeEnabled();
+    expect(button).toHaveAttribute('href', '/dashboard/courses/' + order.course!.code);
+  });
+
+  it('render component with enrollments', () => {
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx:28]:
```suggestion
  it('renders component with enrollments', () => {
```

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+        },
+      ],
+    };
+
+    // const enrollment: Enrollment = JoanieEnrollmentFactory.seed(3).generate();
+    // enrollment.course_run = {
+    //   ...enrollment.course_run,
+    //   end: '2039-10-01T01:23:37+00:00',
+    //   start: '2022-09-09T12:02:44+00:00',
+    // };
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx:60]:
```suggestion
```

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+describe('<DashboardItemOrder/>', () => {
+  it('render component without enrollments', () => {
+    const order: Order = OrderFactory.generate();
+
+    render(
+      <IntlProvider locale="en">
+        <DashboardItemOrder order={order} />
+      </IntlProvider>,
+    );
+    screen.getByText(order.course!.title);
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx:22]:
```suggestion
    screen.getByRole('heading', {level: 5, name: order.course!.title});
```

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+    const button = screen.getByTestId('dashboard-item-order__button');
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx:23]:
```suggestion
    const button = screen.getByRole('link', { name: 'ACCESS COURSE' })
```

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx [hunk 30]
```diff
@@ -0,0 +1,105 @@
...
+            new Date(order.enrollments[0].course_run.start),
+          ),
+      );
+    }
+
+    // Second enrollment.
+    {
+      const element = screen.getByText(order.enrollments[1].course_run.course!.title);
+      const container = element.closest('[data-testid=dashboard-sub-item]') as HTMLElement;
+      // const container = element.closest('[data-testid=dashboard-sub-item]') as HTMLElement;
```
Comment (jbpenrath) [inline src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx:92]:
```suggestion
```

File: src/frontend/js/utils/useDateFormat.tsx [hunk 31]
```diff
@@ -1,5 +1,11 @@
 import { type FormatDateOptions, useIntl } from 'react-intl';
 
+export const FORMAT_DATE_OPTIONS: FormatDateOptions = {
```
Comment (jbpenrath) [inline src/frontend/js/utils/useDateFormat.tsx:3]: What do you think ?
```suggestion
export const DEFAULT_DATE_FORMAT: FormatDateOptions = {
```

Comment (NathanVss) [reply src/frontend/js/utils/useDateFormat.tsx]: Seems legit :) I love putting "type-hint" in names 🥸

File: src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx [hunk 32]
```diff
@@ -0,0 +1,97 @@
...
+      ],
+    };
+
+    render(
+      <IntlProvider locale="en">
+        <DashboardItemOrder order={order} />
+      </IntlProvider>,
+    );
+    screen.getByText(order.course!.title);
+    const button = screen.getByTestId('dashboard-item-order__button');
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/components/DashboardItem/DashboardItemOrder.spec.tsx:61]: Nailed it :')
